Firestore/core/src/firebase/firestore/util/async_queue.cc

```diff
@@ -0,0 +1,314 @@
+// interface: accepts an arbitrary invocable object in place of an Objective-C
+// block.
+template <typename Work>
+void DispatchAsync(dispatch_queue_t queue, Work&& work) {
+  DoDispatch(queue, dispatch_async_f, std::forward<Work>(work));
+}
+
+// Similar to DispatchAsync but wraps dispatch_sync_f.
+template <typename Work>
+void DispatchSync(dispatch_queue_t queue, Work&& work) {
+  DoDispatch(queue, dispatch_sync_f, std::forward<Work>(work));
+}
+
+}  // namespace
+
+namespace detail {
+
+// DelayedOperationImpl contains the logic of scheduling a delayed operation.
+//
+// An instance of this class exists until it's run, which allows it to schedule
+// itself for delayed execution without worrying about lifetime issues.
+//
+// AsyncQueue holds a shared_ptr to the instance, while DelayedOperation
+// handle returned to the code using AsyncQueue holds a weak_ptr.
+// Consequently, DelayedOperation is always valid in the sense that it's
+// always safe to use, but the lifetime of DelayedOperationImpl only depends
+// on the AsyncQueue. AsyncQueue never removes delayed operations on its
+// own; only DelayedOperationImpl itself triggers its removal from the queue
+// in its HandleDelayElapsed method.
+//
+// It is impossible to actually cancel work scheduled with libdispatch; to work
+// around this, DelayedOperationImpl emulates cancelation by turning itself
+// into a no-op. Under the hood, even a "canceled" DelayedOperationImpl will
+// still run, and consequently the instance will still be alive until it's run.
+
+class DelayedOperationImpl {
+ public:
+  DelayedOperationImpl(AsyncQueue* queue,
+                       TimerId timer_id,
+                       AsyncQueue::Milliseconds delay,
+                       AsyncQueue::Operation&& operation)
+      : queue_{queue},
+        timer_id_{timer_id},
+        target_time_{delay},
+        operation_{std::move(operation)} {
+    Start(delay);
+  }
+
+  void Cancel() {
+    queue_->VerifyIsCurrentQueue();
+    done_ = true;
+  }
+
+  void SkipDelay() {
+    queue_->EnqueueAllowingSameQueue([this] { HandleDelayElapsed(); });
+  }
+
+  TimerId timer_id() const {
+    return timer_id_;
+  }
+
+  bool operator<(const DelayedOperationImpl& rhs) const {
+    return target_time_ < rhs.target_time_;
+  }
+
+ private:
+  void Start(const AsyncQueue::Milliseconds delay) {
+    namespace chr = std::chrono;
+    const dispatch_time_t delay_ns = dispatch_time(
+        DISPATCH_TIME_NOW, chr::duration_cast<chr::nanoseconds>(delay).count());
+    dispatch_after_f(
+        delay_ns, queue_->dispatch_queue(), this, [](void* const raw_self) {
+          const auto self = static_cast<DelayedOperationImpl*>(raw_self);
+          self->queue_->EnterCheckedOperation(
+              [self] { self->HandleDelayElapsed(); });
+        });
+  }
+
+  void HandleDelayElapsed() {
+    queue_->VerifyIsCurrentQueue();
+    if (!done_) {
+      done_ = true;
```

Comment:
Could there be race case that leads operation being run twice? I guess not if we can guarantee that a DelayedOperationImpl cannot be put into two different dispatch-queue ever.